development/gcbuild/main.go

```diff
@@ -0,0 +1,340 @@
+	Name string   `yaml:"name"`
+	Args []string `yaml:"args"`
+}
+
+type options struct {
+	buildDir      string
+	configFile    string
+	variantsFile  string
+	variant       string
+	logDir        string
+	devRegistry   string
+	project       string
+	stagingBucket string
+	logsBucket    string
+	silent        bool
+	isCI          bool
+}
+
+// parseVariable returns a gcloud compatible substitution option.
+// Keys are set to upper-case and prefix "_" is set if not present.
+func parseVariable(key, value string) string {
+	k := strings.TrimSpace(strings.ToUpper(key))
+	if !strings.HasPrefix(k, "_") {
+		k = "_" + k
+	}
+	return k + "=" + strings.TrimSpace(value)
+}
+
+func getSHAFromGit() (string, error) {
+	cmd := exec.Command("git", "rev-parse", "head")
+	out, err := cmd.Output()
+	if err != nil {
+		return "", err
+	}
+	validTagRegexp, err := regexp.Compile("[^-_.a-zA-Z0-9]+")
+	if err != nil {
+		return "", err
+	}
+	sanitizedOutput := validTagRegexp.ReplaceAllString(string(out), "")
+	return sanitizedOutput, nil
+}
+
+// run prepares command execution and handles gathering logs to file
+func run(o options, name, repo, tag string, subs map[string]string) error {
+	var s []string
+
+	s = append(s,
+		parseVariable("_TAG", tag),
+		parseVariable("_REPOSITORY", repo),
+	)
+
+	// parse additional substitutions
+	for k, v := range subs {
+		s = append(s, parseVariable(k, v))
+	}
+	args := []string{
+		"builds", "submit",
+		"--config", o.configFile,
+		"--substitutions", strings.Join(s, ","),
+	}
+
+	if o.project != "" {
+		args = append(args, "--project", o.project)
+	}
+	// TODO (@Ressetkk): custom staging bucket implementation and re-using source code in builds
+	//if o.stagingBucket != "" {
+	//
+	//}
+
+	cmd := exec.Command("gcloud", args...)
+
+	var outw []io.Writer
+	var errw []io.Writer
+
+	if !o.silent {
+		outw = append(outw, os.Stdout)
+		errw = append(errw, os.Stderr)
+	}
+
+	f, err := os.Create(filepath.Join(o.logDir, strings.TrimSpace("build_"+strings.TrimSpace(name)+".log")))
+	if err != nil {
+		return fmt.Errorf("could not create log file: %w", err)
+	}
+
+	outw = append(outw, f)
+	errw = append(errw, f)
+
+	cmd.Stdout = io.MultiWriter(outw...)
+	cmd.Stderr = io.MultiWriter(errw...)
+
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+	return nil
+}
+
+type variants map[string]map[string]string
+
+func getVariants(o options) (variants, error) {
+	var v variants
+	b, err := ioutil.ReadFile(o.variantsFile)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+		// variant file not found, skipping
+		return nil, nil
+	}
+	if err := yaml.UnmarshalStrict(b, &v); err != nil {
+		return nil, err
+	}
+	if o.variant != "" {
+		va, ok := v[o.variant]
+		if !ok {
+			return nil, fmt.Errorf("requested variant '%s', but it's not present in variants.yaml file", o.variant)
+		}
+		return variants{o.variant: va}, nil
+	}
+	return v, nil
+}
+
+func getImageNames(repo, tag, variant string, images []string) string {
```

Comment:
What it the purpose of this function? What it's doing?

Reply:
it gets the images from `images` field, replaces the subs variables with actual values and returns aggregated string with a list of images built